### PR TITLE
[BEAM-4543] Python Datastore IO using google-cloud-datastore

### PR DESCRIPTION
**Pull request by udim**
Structure:
apache_beam/io/gcp/datastore/v1 - existing code using googledatastore
apache_beam/io/gcp/datastore/v1new - new code using
  google-cloud-datastore, largely based on v1

Migration path:
Both googledatastore (v1) and google-cloud-datastore (v1new) packages
are listed in the dependencies, but only one can be used in the same
process, due to protobuf module conflicts. Thus, tests for the new and
old clients must be explicitly run seperately (see tox.ini and
build.gradle).
Once the code is deemed stable enough and at least one Beam version is
released that supports both, the old v1 code will be removed along with
the googledatastore dep.

Changes compared to the old client:
- Transforms now use types.{Query,Key,Entity} classes instead of
protobufs.
For more, please see
https://docs.google.com/document/d/1sL9p7NE5Z0p-5SB5uwpxWrddj_UCESKSrsvDTWNKqb4/edit#bookmark=id.hwxlz1oe7v5e
(and please comment if you have additional questions).

Notes:
- For mutations, the new client (google-cloud-datastore) does have
mechanisms for retries (not enabled), but I've opted to keep the
existing mechanism, which does retries with exponential backoff and
adaptive throttling (dynamic sizing of commit batches).
- Integration test does query, upsert, and delete. Should be running on
{py2,py3} x {direct_runner,dataflow_runner}.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by udim** [comment]
R: @chamikaramj

**Comment by udim** [comment]
run python postcommit

**Comment by udim** [comment]
run python postcommit

**Comment by udim** [comment]
run python postcommit

File: sdks/python/apache_beam/io/gcp/datastore/v1new/query_splitter.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
```
[the remark this answers is not in the document]
Comment (udim) [reply sdks/python/apache_beam/io/gcp/datastore/v1new/query_splitter.py]: Yeah it is different. I've refactored the unit tests though to reuse v1 tests as much as possible.

**Comment by udim** [comment]
run python postcommit

**Comment by udim** [comment]
PTAL

**Comment by udim** [comment]
run python postcommit

File: sdks/python/apache_beam/io/gcp/datastore/v1new/datastore_write_it_test.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                  'num_entities': self.NUM_ENTITIES,
+                  'on_success_matcher': all_of(*pipeline_verifiers)}
+    if limit is not None:
+      extra_opts['limit'] = limit
+
+    datastore_write_it_pipeline.run(test_pipeline.get_full_options_as_args(
+        **extra_opts))
+
+  @attr('IT')
+  def test_datastore_write_limit(self):
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/datastore/v1new/datastore_write_it_test.py]: Can you add a IT for read as well ?

File: sdks/python/setup.py [hunk 3]
```diff
@@ -137,11 +137,14 @@ def get_version():
     ]
 
 GCP_REQUIREMENTS = [
+    'cachetools>=3.1.0,<4',
     # google-apitools 0.5.23 and above has important Python 3 supports.
     'google-apitools>=0.5.26,<0.5.27',
-    'proto-google-cloud-datastore-v1>=0.90.0,<=0.90.4',
```
Comment (chamikaramj) [inline sdks/python/setup.py]: Don't we need proto-google-cloud-datastore-v1 for v1/datastoreio ?

File: sdks/python/setup.py [hunk 4]
```diff
@@ -137,11 +137,14 @@ def get_version():
...
-    # [BEAM-4543] Datastore IO is not supported in Python 3.
+    # [BEAM-4543] googledatastore is not supported in Python 3.
```
Comment (chamikaramj) [inline sdks/python/setup.py]: Can both google-cloud-datastore and googledatastore installed in the same env without conflicts ?

Comment (udim) [reply sdks/python/setup.py]: Yes, having both installed is a large contributor to the complexity of this PR.
You can't import both in the same process.

File: sdks/python/setup.py (hunk 3 above)
Comment (udim) [reply sdks/python/setup.py]: Yes. it's still listed in this file below.

File: sdks/python/apache_beam/io/gcp/datastore/v1new/datastore_write_it_test.py (hunk 2 above)
Comment (udim) [reply sdks/python/apache_beam/io/gcp/datastore/v1new/datastore_write_it_test.py]: The name is misleading (inherited from the v1 test). It does read, write, and delete.

File: sdks/python/apache_beam/io/gcp/datastore/v1new/query_splitter_test.py [hunk 5]
```diff
@@ -173,5 +163,9 @@ def test_client_key_sort_key(self):
     self.assertEqual(expected_sort, keys)
 
 
+# Hide base class from collection by nose.
+del QuerySplitterTestBase
```
[the remark this answers is not in the document]
Comment (udim) [reply sdks/python/apache_beam/io/gcp/datastore/v1new/query_splitter_test.py]: Yes, they do. This doesn't delete the base class, only the name for it .

**Comment by udim** [comment]
run python postcommit

File: sdks/python/apache_beam/io/gcp/datastore/v1/datastoreio.py [hunk 6]
```diff
@@ -45,6 +50,10 @@
 try:
   from google.cloud.proto.datastore.v1 import datastore_pb2
   from googledatastore import helper as datastore_helper
+  logging.warning(
+      'Using deprecated Datastore client.\n'
+      'This client will be removed in the next Beam major release.\n'
```
[the remark this answers is not in the document]
Comment (udim) [reply sdks/python/apache_beam/io/gcp/datastore/v1/datastoreio.py]: done

**Comment by chamikaramj** [comment]
LGTM. Thanks Udi.

Will merge after tests pass. Also please squash commits as needed.

**Comment by udim** [comment]
Latest force push squashed commits and added import skipIf()s to both `datastore_write_it_test.py`s.

**Comment by udim** [comment]
run python postcommit

**Comment by chamikaramj** [comment]
Run Python PostCommit

**Comment by jacobg** [comment]
I think there's a bug in the new datastore implementation, where it's skipping entities. Here's the JIRA issue I just opened:

https://issues.apache.org/jira/browse/BEAM-7608

**Pull request merged**